### DISCUSSION TITLE
Virtual IP pool lease time and reassignment

### DISCUSSION THREAD

**Discussion by JanSwan** [strongswan/strongswan#1378] [Q&A] [answered]
Hi

I'm using virtual pool for assigning IP-s to Android RW clients and noticed that user gets assigned new IP with every connection (tried it by setting phone to flight mode and exiting flight mode again). 

swanctl --list-pools -l
site                  10.199.75.161                       16 / 0 / 62
  10.199.75.161                  online   'K48@site'
  10.199.75.162                  online   'K48@site'
  10.199.75.165                  online   'K48@site'
  10.199.75.164                  online   'K48@site'
  10.199.75.168                  online   'K48@site'
  10.199.75.169                  online   'K48@site'
  10.199.75.170                  online   'K48@site'
  10.199.75.171                  online   'K48@site'
  10.199.75.172                  online   'K48@site'
  10.199.75.173                  online   'K48@site'
  10.199.75.174                  online   'K48@site'
  10.199.75.175                  online   'K48@site'
  10.199.75.176                  online   'K48@site'
  10.199.75.177                  online   'K48@site'
  10.199.75.178                  online   'K48@site'
  10.199.75.179                  online   'K48@site'

This depletes my pool. Is it possible to :
1. reassign same IP to same ID (i'm using same PSK, but different ID-s) ?
2. modify lease parameters (how long is it kept online - and offline, before ip address can be reassigned) ?

In connection conf i have
pools = site
remote {
         auth = psk
         id=*@site
      }

Best

Jaanus

**Comment by JanSwan** [comment]
PS! I'm aware of possibility to create single IP pools or connections with ".vips=X.X.X.X" for every ID, but i'm looking for more "modern" solution, that doesn't include creation of 62 separate connections.

**Comment by tobiasbrunner** [comment, marked as answer]
> 1. reassign same IP to same ID (i'm using same PSK, but different ID-s) ?

That's a bad idea! Every user can impersonate any other user by simply changing the ID and even impersonate the server and act as man-in-the-middle as everybody knows that PSK. At least use individual PSKs. But I'd recommend you use a certificate to authenticate the server and EAP with different username/password combinations for the clients. That also prevents active attacks on the PSKs as the server is first authenticated before the clients send their password hashes.

IPs are assigned per IKE_SA and per identity. So if multiple IKE_SAs with the same identity are established concurrently, there will be multiple IPs assigned (the in-memory address pool does not reassign active leases unless in reauthentication scenarios where the client connects from the same IP address and port). To prevent clients from creating multiple IKE_SAs, you may configure the `unique` setting for the connection.

There are also address pool implementations that allow assigning static leases per identity (e.g. via SQL database, RADIUS or DHCP servers, see [here](https://docs.strongswan.org/docs/5.9/features/vip.html#_database_backend) for details).


> 2. modify lease parameters (how long is it kept online - and offline, before ip address can be reassigned) ?

That's not a propery of the address pool but depends on the currently established IKE_SAs. If the client doesn't close previous IKE_SAs actively, the server has to do that by triggering DPD exchanges and checking if the client still responds. Note that configuring a too aggressive DPD interval could be problematic, as the server might close active IKE_SAs the client might still use just because the client is not reachable for a while (e.g. because it is roaming between networks or saving power). But that might depend on the client (for instance strongSwan's Android app will keep the same IKE_SA alive even when roaming or being offline for some time via MOBIKE, but if you are using the built-in client that might be different).

**Reply by JanSwan** [reply]
Hi Tobias.

Thanks for answer. 
I'm pushing VPN profile to the phones via MDM, so users don't know the psk, neither can they change the ID (device name) from device. At least not from GUI. I agree with you that RSA (certificates) auth would be more secure, but unfortunately MDM doesn't support IKEv2 RSA option and the MDM dev answered that due to KNOX API declared eol soon, it's currently pointless to add support for it. 
All my IOS devices use certificate auth and DHCP server on the background, so i know it works :) Those Android devices have specific function and I didn't want them to use same DHCP server pool - that's the only reason I chose Virtual IP-s.
I was also aware of "unique" but aparently the manual I read was for ipsec, where unique=yes was default behaviour :) I just found out that with swanctl default=no - my bad. 

So, setting unique=replace makes things much better. The lease still stays in table, but gets offline status much faster, so hopefully it will solve my problem, if not i'll implement any other proposed solution (SQL,DHCP etc). 

Logs to show "unique" works, lease is not deleted together with SA, but it is marked offline(reusable) in matter of few mins, so works for me.

Oct 27 15:50:27 StronSWAN charon-systemd[13312]: 06[IKE] deleting duplicate IKE_SA for peer 'K43@site' due to uniqueness policy
Oct 27 15:50:27 StronSWAN charon-systemd[13312]: 06[IKE] deleting IKE_SA psk_auth_site[13] between 10.199.75.2[site.site.dom]...10.199.78.221[K43@site]
Oct 27 15:50:27 StronSWAN charon-systemd[13312]: 06[IKE] sending DELETE for IKE_SA psk_auth_site[13]
Oct 27 15:50:27 StronSWAN charon-systemd[13312]: 06[IKE] peer requested virtual IP %any
Oct 27 15:50:27 StronSWAN charon-systemd[13312]: 06[CFG] assigning new lease to 'K43@site'
Oct 27 15:50:27 StronSWAN charon-systemd[13312]: 06[IKE] assigning virtual IP 10.199.75.166 to peer 'K43@site'

Oct 27 15:52:40 StronSWAN charon-systemd[13312]: 10[IKE] deleting duplicate IKE_SA for peer 'K43@site' due to uniqueness policy
Oct 27 15:52:40 StronSWAN charon-systemd[13312]: 10[IKE] deleting IKE_SA psk_auth_site[19] between 10.199.75.2[site.site.dom]...10.199.78.221[K43@site]
Oct 27 15:52:40 StronSWAN charon-systemd[13312]: 10[IKE] sending DELETE for IKE_SA psk_auth_site[19]
Oct 27 15:52:40 StronSWAN charon-systemd[13312]: 10[IKE] peer requested virtual IP %any
Oct 27 15:52:40 StronSWAN charon-systemd[13312]: 10[CFG] reassigning offline lease to 'K43@site'
Oct 27 15:52:40 StronSWAN charon-systemd[13312]: 10[IKE] assigning virtual IP 10.199.75.163 to peer 'K43@site'

Oct 27 15:55:06 StronSWAN charon-systemd[13312]: 15[IKE] deleting duplicate IKE_SA for peer 'K43@site' due to uniqueness policy
Oct 27 15:55:06 StronSWAN charon-systemd[13312]: 15[IKE] deleting IKE_SA psk_auth_site[27] between 10.199.75.2[site.site.dom]...10.199.78.221[K43@site]
Oct 27 15:55:06 StronSWAN charon-systemd[13312]: 15[IKE] sending DELETE for IKE_SA psk_auth_site[27]
Oct 27 15:55:06 StronSWAN charon-systemd[13312]: 15[IKE] peer requested virtual IP %any
Oct 27 15:55:06 StronSWAN charon-systemd[13312]: 15[CFG] reassigning offline lease to 'K43@site'
Oct 27 15:55:06 StronSWAN charon-systemd[13312]: 15[IKE] assigning virtual IP 10.199.75.164 to peer 'K43@site'


swanctl --list-pools -l
 
  10.199.75.163                  online   'K43@site'
  10.199.75.164                  online   'K43@site'
  10.199.75.166                  offline  'K43@site'

Again thanks for answer, it was really helpful.

Best

Jaanus
